packages/@aws-cdk/aws-lambda/lib/function-base.ts

```diff
@@ -431,6 +435,35 @@ export abstract class FunctionBase extends Resource implements IFunction, ec2.IC
     throw new Error(`Invalid principal type for Lambda permission statement: ${principal.constructor.name}. ` +
       'Supported: AccountPrincipal, ArnPrincipal, ServicePrincipal');
   }
+
+  private parseConditions(principal: iam.IPrincipal): { sourceAccount: string, sourceArn: string } | null {
+    if ('conditions' in principal) {
```

Comment:
I would rewrite this to a type testing function and then use that here:

```ts
function isPrincipalWithConditions(x: iam.IPrincipal): x is iam.PrincipalWithConditions {
  return 'conditions' in principal;
}
```

Saves you a type cast afterwards and is very clear in what it does.